docs/RFCS/20200401_alter_column_type.md

```diff
@@ -0,0 +1,87 @@
+- [Problems](#Problems)
+- [Other considerations](#Other-considerations)
+- [Alternatives](#alternatives)
+
+# Summary
+Support changing type of a column. (For nontrivial type conversions using a cast).
+
+# Motivation
+Fairly common use case to change a columns data type, especially in ORMs. Highly requested feature.
+
+# High-level Summary
+The idea for a column conversion requiring a cast is to create a new computed column by applying the cast/function to the original column. For each index that contains that column, create a new index that uses all the columns in i but replaces the old column being converted with the new computed column. Swap the old index with the new index, swap the old column with the new column. Then drop the old columns and indexes.
+
+# Detailed Explanation
+## Converting a column that is not part of an index (simple case)
+If were trying to convert column **c** to type **t**
+1. Create a new hidden computed column c' using c::t'
+2. Enqueue a column swap mutation, for c and c'
+    - Make c a computed column of c'
+        - Only works if we have an inverse - this makes alter column type using an expression extremely tricky
+    - Make c hidden, make c' not hidden
+    - Swap the names of c and c'
+    - Swap positions of c and c' in TableDescriptor.Columns
+        - This is where physical order of columns is determined
+    - Enqueue drop for c'
+    - Note that this current process leaves the ColumnIDs and thus the ordinal positions of the columns out of order. Explained in out of order column ids section.
+
+## If the column is part of an index (not part of primary key)
+1. Create a new hidden computed column c' using c::t'
+2. For each index c' is part of, create a new index i' as a duplicate of i but replace each reference to c with c'.
+3. Enqueue index swap mutations.
+4. Enqueue column swap mutation. 
+5. Enqueue drops for the old column and old indexes. 
+
+## If the column is part of the primary key
+1. Create a new hidden computed column c' using c::t'
+2. Make a copy of the primary key pk and replace each reference to c with c'
+    - pk' must be encoded as a primary index, cannot switch secondary index to primary index
+3. Each index i that references pk, create an index i' that references pk'
+4. Enqueue primary index swap with pk and pk', and index swaps with i and i'. 
+5. Swap c' with c. 
+6. Enqueue drops for the old column and old indexes including the primary key.
+
+
+
+# Problems
+## Out of order ColumnIDs/OrdinalPositions/attnum
+When performing a column swap, ColumnIDs cannot be swapped due to encodings. For the current listed method of performing a swap by adding a new column and dropping the old one, the ColumnIDs will be "out of order". This results in places that currently depend on ColumnIDs to also be out of order. For example in information_schema.columns, ordinal_position is ColumnID. (Similarly attnum and adnum in pg_catalog also correpsond to ColumnID).
```

Comment:
"Due to column IDs being a part of the value encoding for columns, we cannot swap the column IDs of the new and old column" ( or something like that).

Reply:
Added